### PR TITLE
Change architecture to permit usage of error builder with another serializer than marshmallow

### PR DESCRIPTION
**Pull request by buxx**
Related #118

### CODE REVIEW

File: hapic/decorator.py [hunk 1]
```diff
@@ -62,6 +64,46 @@ def get_doc_string(self) -> str:
...
+        self,
+        context: typing.Union[
+            ContextInterface, typing.Callable[[], ContextInterface]
+        ],
+        processor_factory: typing.Callable[
+            [typing.Optional[TYPE_SCHEMA]], Processor
+        ],
+        error_http_code: HTTPStatus = HTTPStatus.BAD_REQUEST,
+        default_http_code: HTTPStatus = HTTPStatus.OK,
+    ) -> None:
```
Comment (inkhey) [inline hapic/decorator.py]: docstring maybe ?

File: hapic/context.py [hunk 2]
```diff
@@ -30,6 +33,17 @@ class ContextInterface(object):
     def get_request_parameters(self, *args, **kwargs) -> RequestParameters:
         raise NotImplementedError()
 
+    def set_processor_class(
+        self, processor_class: typing.Type[Processor]
+    ) -> None:
+        """
+        Must update processor class associated to this context.
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/context.py]: Description changes.

File: hapic/context.py [hunk 3]
```diff
@@ -154,10 +168,25 @@ def __init__(
 
 
 class BaseContext(ContextInterface):
+    def __init__(
+        self, processor_class: typing.Optional[typing.Type[Processor]] = None
+    ) -> None:
+        self._processor_class = processor_class
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/context.py]: Documentation added about that.

File: hapic/decorator.py [hunk 4]
```diff
@@ -645,24 +672,26 @@ def _build_error_response(self, exc: Exception) -> typing.Any:
...
             exc, include_traceback=self.context.is_debug()
         )
-
+        processor = self._processor_factory(self.error_builder.get_schema())
         # Check error format
-        dumped = self.error_builder.dump(response_content).data
-        unmarshall = self.error_builder.load(dumped)
-        if unmarshall.errors:
+        try:
+            dumped = processor.validate_error(response_content)
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/decorator.py]: Architecture changed

File: hapic/ext/aiohttp/context.py [hunk 5]
```diff
@@ -107,7 +109,21 @@ def __init__(
...
-                        dumped = error_builder.dump(error_body).data
+                        processor = self._processor_class(
+                            error_builder.get_schema()
+                        )
+
+                        # TODO BS 2018-12-07: error validation is some
+                        # duplicated code. It could be an refactored code.
+                        # See #123.
+                        try:
+                            dumped = processor.validate_error(error_body)
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/ext/aiohttp/context.py]: Architecture changed.

File: hapic/ext/aiohttp/context.py [hunk 6]
```diff
@@ -145,16 +161,20 @@ def get_validation_error_response(
...
 
         # Check error
-        dumped = error_builder.dump(error_content).data
-        unmarshall = error_builder.load(dumped)
-        if unmarshall.errors:
+        # TODO BS 2018-12-07: error validation is some
+        # duplicated code. It could be an refactored code.
+        # See #123.
+        try:
+            dumped = processor.validate_error(error_content)
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/ext/aiohttp/context.py]: Architecture changed

File: hapic/ext/flask/context.py [hunk 7]
```diff
@@ -72,17 +74,23 @@ def get_validation_error_response(
...
         # Check error
-        dumped = self.default_error_builder.dump(error_content).data
-        unmarshall = self.default_error_builder.load(dumped)
-
-        if unmarshall.errors:
+        # TODO BS 2018-12-07: error validation is some
+        # duplicated code. It could be an refactored code.
+        # See #123.
+        try:
+            dumped = processor.validate_error(error_content)
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/ext/flask/context.py]: architecture changed

File: hapic/decorator.py (hunk 1 above)
Comment (buxx) [reply hapic/decorator.py]: done

File: hapic/processor/serpyco.py [hunk 8]
```diff
@@ -229,6 +233,30 @@ def load_input(self, input_data: typing.Any) -> typing.Any:
...
 
+    def validate_error(self, data_to_validate: typing.Any) -> typing.Any:
+        """
+        Use schema to validate givan data.
+        Raise ProcessValidationError in case of validation error
+        :param data_to_validate: data to validate
+        :return validated and potentially updated data (like with
+        default values)
+        """
+        # TODO BS 2018-11-26: This is correct to do this every times ?
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/processor/serpyco.py]: myself. I have changed it by a note about the why of it.

File: hapic/ext/flask/context.py [hunk 9]
```diff
@@ -72,17 +74,23 @@ def get_validation_error_response(
...
+        )
 
         # Check error
-        dumped = self.default_error_builder.dump(error_content).data
-        unmarshall = self.default_error_builder.load(dumped)
-
-        if unmarshall.errors:
+        # TODO BS 2018-12-07: error validation is some
+        # duplicated code. It could be an refactored code.
+        # See #123.
```
[the remark this answers is not in the document]
Comment (buxx) [reply hapic/ext/flask/context.py]: done

**Comment by buxx** [comment]
All green and code re-architectured ! :)

**Pull request merged**